File: Java/medium/decodeXORedPermutation.java
import java.util.*;

class Solution {
    public static int[] decode(int[] encoded) {
        int[] ans = new int[encoded.length + 1];
        int total = 0;
        int excludeFirst = 0;

        for(int i = 1; i <= encoded.length + 1; i++){
            total ^= i;
        }

        for(int i = 1; i < encoded.length; i+= 2){
            excludeFirst ^= encoded[i];
        }

        ans[0] = total ^ excludeFirst;
        for(int i = 0; i < encoded.length; i++){
            ans[i + 1] = encoded[i] ^ ans[i];
        }
        return ans;
    }

    // encoded = [3, 1]
    // decoded = [1, 2, 3]      
    
    // encoded = [6, 5, 4, 6]   
    // decoded = [2, 4, 1, 5, 3]

    public static void main(String[] args){
        int[][] testSet = new int[][]{{3,1}, {6,5,4,6}};

        for(int[] test : testSet){
            System.out.println("encoded = " + Arrays.toString(test));
            int[] decoded = decode(test);
            System.out.println("decoded = " + Arrays.toString(decoded) + '\n');
        }
    }
}

File: CPP/medium/decodeXORedPermutation.cpp
#include <vector>
#include <iostream>

std::vector<int> decode(std::vector<int>& encoded) {
    int odd = 0;
    int even = 0;
    for(int i = 1; i <= encoded.size() +1; i++){
        odd ^= i;
    }
    for(int i = 1; i < encoded.size(); i+=2){
        even ^= encoded[i];
    }
    std::vector<int> ans = {odd ^ even};
    for(int i = 1; i < encoded.size() + 1; i++){
        ans.push_back(encoded[i - 1] ^ ans[i - 1]);
    }
    return ans;
}

// encoded = {3,1}
// decoded = {1,2,3}

// encoded = {6,5,4,6}
// decoded = {2,4,1,5,3}

int main(){
    std::vector<std::vector<int>> testSet = {
        {3,1}, {6,5,4,6}
    };
    for(std::vector<int> test : testSet){
        std::cout << "encoded = {";
        for(int i = 0; i < test.size(); i++){
            std::cout << test[i] << (i == (test.size() - 1)? "}\n":",");
        }
        std::vector<int> decoded = decode(test);
        std::cout << "decoded = {";
        for(int i = 0; i < decoded.size(); i++){
            std::cout << decoded[i] << (i == (decoded.size() - 1)? "}\n":",");
        }
        std::cout << std::endl;
    }
    return 0;
}